apps/smart-contracts/token/test/PPO.test.ts

```diff
@@ -445,6 +449,402 @@ describe('=> PPO', () => {
     })
   })
 
+  describe('# transferFromWithPermit', () => {
+    let currentTime: number
+    let deadline: number
+    const BLOCK_DURATION_IN_SECONDS = 15
+    beforeEach(async () => {
+      await setupPPOAndFakeTransferHook()
+      await ppo.connect(owner).mint(user1.address, 10)
+      currentTime = await getLastTimestamp(ethers.provider)
+      deadline = currentTime + BLOCK_DURATION_IN_SECONDS
+    })
+
+    it('reverts if deadline expired', async () => {
+      const timeAfterDeadlineExpiration = deadline + BLOCK_DURATION_IN_SECONDS
+      const user1PPOBalanceBefore = await ppo.balanceOf(user1.address)
+      const { v, r, s } = await getPermitSignature(
+        user1,
+        ppo,
+        spender.address,
+        user1PPOBalanceBefore,
+        deadline
+      )
+      // eslint-disable-next-line @typescript-eslint/no-explicit-any
+      await setNextTimestamp(ethers.provider as any, timeAfterDeadlineExpiration)
+
+      await expect(
+        ppo
+          .connect(spender)
+          .transferFromWithPermit(
+            user1.address,
+            user2.address,
```

Comment:
```suggestion
            user2.address,
```
This is a bad test because you've generating a permit signature for `spender.address` to spend, yet passed in `user2.address`, so it would fail even with a good deadline.

This is why i keep emphasizing you need to write your tests in a way that makes it clear what inputs are changing if your tests change inputs alot, especially with a function that has this many parameters and ones that aren't obvious. Like `timeAfterDeadlineExpiration` is fine because its very clear that this time is past the deadline since we can see it is `deadline + BLOCK_DURATION_IN_SECONDS`, so there is also no need for an expect.

It's harder with an input like an account because you are juggling 3 accounts around all the time.

Reply:
I don't see how you concluded that the test will fail even with a good deadline when it doesn't (the test does not revert if you pass in `deadline` to `setNextTimestamp`. 

As per the current implementation, a user (`user1`) approves an account to spend its tokens and then the approved account can transfer the tokens to a recipient (other than user1 also), here the approved account is `spender` hence the spender can now transfer those tokens to another user (`user2`)

Reply:
Ah yes this is my confusion, I forgot that the `spender` can send to another address, this is my bad.